lib/frontend/src/lib.rs

```diff
@@ -1,12 +1,46 @@
 //! Cranelift IR builder library.
 //!
 //! Provides a straightforward way to create a Cranelift IR function and fill it with instructions
-//! translated from another language. Contains an SSA construction module that lets you translate
-//! your non-SSA variables into SSA Cranelift IR values via `use_var` and `def_var` calls.
+//! corresponding to your source program written in another language.
 //!
 //! To get started, create an [`FunctionBuilderContext`](struct.FunctionBuilderContext.html) and
 //! pass it as an argument to a [`FunctionBuilder`](struct.FunctionBuilder.html).
 //!
+//! # Mutable variables and Cranelift IR values
+//!
+//! The most interesting feature of this API is that it provides a single way to deal with all your
+//! variable problems. Indeed, the [`FunctionBuilder`](struct.FunctionBuilder.html) struct has a
+//! type parameter `Variable` that should be instantiated with the type of your source language
+//! variables. Then, through calling the functions
+//! [`declare_var`](struct.FunctionBuilder.html#method.declare_var),
+//! [`def_var`](struct.FunctionBuilder.html#method.def_var) and
+//! [`use_var`](struct.FunctionBuilder.html#method.use_var), the
+//! [`FunctionBuilder`](struct.FunctionBuilder.html) will create for you all the Cranelift IR
+//! values corresponding to your variables.
+//!
+//! This API has been designed to help you translate your mutable variables into
+//! [`SSA`](https://en.wikipedia.org/wiki/Static_single_assignment_form) form.
+//! [`use_var`](struct.FunctionBuilder.html#method.use_var) will returns the Cranelift IR value
+//! that corresponds to your mutable variable at a precise point in the program. However, you know
+//! beforehand that one of your variables is defined only once, for instance if it is the result
+//! of an intermediate expression in an expression-based language, then you can translate it
+//! directly by the Cranelift IR value returned by the instruction builder. Using the
+//! [`use_var`](struct.FunctionBuilder.html#method.use_var) API for such an immutable variable
+//! would also work but with a slight additional overhead (the SSA algorithm does not know
+//! beforehand if a variable is immutable or not).
+//!
+//!
+//! The moral is that you should use these three functions to handle all your mutable variables, even those
+//! that are not present in the source code but artefacts of the translation. Hence The `Variable` type that you
+//! would pass to [`FunctionBuilder`](struct.FunctionBuilder.html) could look like this
+//!
+//! ```
+//! enum Variable {
+//!     OriginalSourceVariable(String),
+//!     TranslationArtefact(u32)
+//! }
+//! ```
```

Comment:
`FunctionBuilder` requires the `Variable` parameter to implement [`EntityRef`](https://docs.rs/cranelift-entity/0.18.1/cranelift_entity/trait.EntityRef.html), so it doesn't work well to put `String`s or other data in it. And the advice to use source language variables with it only works for source language variable types which are easy to implement `EntityRef` on.

I'm actually pondering whether we should remove the `Variable` type parameter and just use a concrete `Variable` type instead, and require API users to convert their variables into and out of `Variable` manually. That could be a little tedious, but because `Variable`s were already required by the `EntityRef` trait to basically be `usize` wrappers, so it should be straightforward to convert such types into a concrete `Variable` type (which would itself be a trivial `usize` wrapper). I'm interested in what you think about this approach.

Reply:
You're right about the `String` not being a good example. Concerning whether we should make the `Variable` type concrete or not, I agree with you. Looking at the `EntityRef` trait closer, I now realize only aliases of `usize` (or another integer type) can implement it. So it makes no sense to introduce parametricity for something that can only be one type. It will be up to the user to keep a mapping between their program variables and the `usize` index that correspond to it. In that sense I suggest renaming it to `VariableIndex`.